cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -782,6 +941,10 @@ class InputState {
   std::vector<Type::type> key_type_id_;
   // Hasher for key elements
   mutable KeyHasher* key_hasher_;
+  // Owning node
+  AsofJoinNode& node_;
```

Comment:
We want to avoid mutable references, as per the C++ style guide. References should be const, otherwise use a pointer to make it clear at the call site that the argument may be mutated.
```suggestion
  AsofJoinNode* node_;
```